File: 高级数据结构与算法/Python/第3章 排序算法/3.4 快速排序/quick_sort_iteration.py
import stack

def partition(lst, start, end):
    i = start - 1
    pivot = lst[end]

    for j in range(start, end):
        if lst[j] <= pivot:
            i += 1
            lst[i], lst[j] = lst[j], lst[i]

    lst[i+1], lst[end] = lst[end], lst[i+1]
    return i + 1

def quick_sort(lst, start, end):
    s = stack.Stack()
    s.push(start)
    s.push(end)

    while not s.is_empty():
        right = s.pop()
        left = s.pop()

        index = partition(lst, left, right)
        if index - 1 > left:
            s.push(left)
            s.push(index - 1)
        if index + 1 < right:
            s.push(index + 1)
            s.push(right)

def main():
    lst = [5, 8, 6, 3, 9, 2, 1, 7]
    quick_sort(lst, 0, len(lst)-1)
    print(lst)

if __name__ == "__main__":
    main()

File: 高级数据结构与算法/C/第3章 排序算法/3.4 快速排序/quick_sort_iteration.c
#include <stdio.h>
#include "stack.h"

void showArr(int *arr, int n) {
    for(int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int *arr, int start, int end) {
    int i = start - 1;
    int pivot = arr[end];

    for(int j = start; j < end; j++) {
        if(arr[j] <= pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }

    swap(&arr[i+1], &arr[end]);
    return i + 1;
}

void quickSort(int *arr, int start, int end) {
    Stack *s = initStack(end - start + 1);
    push(s, start);
    push(s, end);

    while(!isEmptyStack(s)) {
        int right = pop(s);
        int left = pop(s);

        int index = partition(arr, left, right);
        if(index - 1 > left) {
            push(s, left);
            push(s, index - 1);
        }
        if(index + 1 < right) {
            push(s, index + 1);
            push(s, right);
        }
    }
}

int main() {
    int arr[] = {5, 8, 6, 3, 9, 2, 1, 7};
    int n = sizeof(arr) / sizeof(arr[0]);
    quickSort(arr, 0, n-1);
    showArr(arr, n);
    return 0;
}